pkg/runner/run_context.go

```diff
@@ -259,7 +271,66 @@ func (rc *RunContext) startJobContainer() common.Executor {
 		ext := container.LinuxContainerEnvironmentExtensions{}
 		binds, mounts := rc.GetBindsAndMounts()
 
+		// specify the network to which the container will connect when `docker create` stage. (like execute command line: docker create --network <networkName> <image>)
+		// if using service containers, will create a new network for the containers.
+		// and it will be removed after at last.
+		networkName := rc.networkName()
+
+		// add service containers
+		for serviceID, spec := range rc.Run.Job().Services {
+			// interpolate env
+			interpolatedEnvs := make(map[string]string, len(spec.Env))
+			for k, v := range spec.Env {
+				interpolatedEnvs[k] = rc.ExprEval.Interpolate(ctx, v)
+			}
+			envs := make([]string, 0, len(interpolatedEnvs))
+			for k, v := range interpolatedEnvs {
+				envs = append(envs, fmt.Sprintf("%s=%s", k, v))
+			}
+			username, password, err = rc.handleServiceCredentials(ctx, spec.Credentials)
+			if err != nil {
+				return fmt.Errorf("failed to handle service %s credentials: %w", serviceID, err)
+			}
+			serviceBinds, serviceMounts := rc.GetServiceBindsAndMounts(spec.Volumes)
+			serviceContainerName := createContainerName(rc.jobContainerName(), serviceID)
+			c := container.NewContainer(&container.NewContainerInput{
+				Name:           serviceContainerName,
+				WorkingDir:     ext.ToContainerPath(rc.Config.Workdir),
+				Image:          spec.Image,
+				Username:       username,
+				Password:       password,
+				Env:            envs,
+				Mounts:         serviceMounts,
+				Binds:          serviceBinds,
+				Stdout:         logWriter,
+				Stderr:         logWriter,
+				Privileged:     rc.Config.Privileged,
+				UsernsMode:     rc.Config.UsernsMode,
+				Platform:       rc.Config.ContainerArchitecture,
+				Options:        spec.Options,
+				NetworkMode:    networkName,
+				NetworkAliases: []string{serviceID},
+			})
+			rc.ServiceContainers = append(rc.ServiceContainers, c)
+		}
+
 		rc.cleanUpJobContainer = func(ctx context.Context) error {
+			if len(rc.ServiceContainers) > 0 {
+				logger.Infof("Cleaning up services for job %s", rc.JobName)
+				if err := rc.stopServiceContainers()(ctx); err != nil {
+					logger.Errorf("Error while cleaning services: %v", err)
+				}
+				if !rc.IsHostEnv(ctx) {
+					// clean network in docker mode only
+					// if using service containers
+					// it means that the network to which containers are connecting is created by `act_runner`,
+					// so, we should remove the network at last.
+					logger.Infof("Cleaning up network for job %s, and network name is: %s", rc.JobName, rc.networkName())
+					if err := container.NewDockerNetworkRemoveExecutor(rc.networkName())(ctx); err != nil {
+						logger.Errorf("Error while cleaning network: %v", err)
```

Comment:
I'm running into:
`Error while cleaning network: Error response from daemon: error while removing network: network act-push-job-2-cf0efa14d6309df9584a42f3b9aaec4bca252b8008d44003059bdc53346e5553-job-2-network id c65ca709ee6811b8c8907bf864aff5b3bb8ad65b0a62b841413d67dce8529ed2 has active endpoints`
when I stop a workflow with ctrl+c
And after that starting the workflow again results in:
`Error: failed to start container: Error response from daemon: network act-push-job-1-b5db9dd46498cf67a2d34f06188bd6188e01956cd86b4dbf3fe53923246b1467-job-1-network is ambiguous (2 matches found on name)`
Here's a simple example workflow to reproduce this behavior: https://github.com/ZauberNerd/act-test/actions/runs/5940389109/workflow

Reply:
I'll take a closer look at this later in the week

Reply:
Feel free to make any changes needed